python/sidecar/references/service.py

```diff
@@ -0,0 +1,209 @@
+import shutil
+from collections import defaultdict
+from uuid import uuid4
+
+import requests
+from requests.exceptions import HTTPError
+from sidecar import shared
+from sidecar.config import logger
+from sidecar.projects import service as projects_service
+from sidecar.references import storage
+from sidecar.references.schemas import IngestStatus, Reference, ReferenceCreate
+from sidecar.shared import chunk_reference
+
+
+def fetch_pdf_to_uploads(
+    url: str, project_id: str, user_id: str, metadata: ReferenceCreate
+) -> ReferenceCreate:
+    """
+    Fetches a PDF from a URL and saves it to the project's uploads directory.
+
+    Parameters
+    ----------
+    url : str
+        The URL of the PDF to fetch.
+    project_id : str
+        The ID of the project to add the reference to.
+    user_id : str
+        The ID of the user who owns the project.
+    metadata : dict
+        The metadata to add to the reference.
+
+    Returns
+    -------
+    ReferenceCreate
+        The metadata of the reference to create.
+    """
+    response = requests.get(url)
+
+    if response.status_code != 200:
+        raise HTTPError(f"Unable to fetch {url} (status code: {response.status_code})")
+
+    if not metadata.source_filename:
+        metadata.source_filename = f"{metadata.title[:200]}.pdf"
```

Comment:
Is `metadata.title[:200]` a way of truncating the title to a max of 200 characters? If so, seems too much. The filename should be pretty opaque for the client code (and server) so I think that would be safer to just use an ID (maybe of the reference) instead of the title, that might also include invalid characters for a url.

Reply:
I can see the point that 200 might be too long though, so I'll shorten this to 100 for now.

I had the thought of showing the ID, but since we show the filename in the file browser on the left side, I thought seeing a random UUID would be a confusing user experience.

We also can't rely on the filename from the URL, since it's often a random ID or even sometimes just `pdf`.

IMO it'd be best if we asked a user to provide a filename when adding the reference from search results. I liken this to when you're adding a file to VS Code or downloading a PDF to your local -- you need to provide a name for it. I think a similar principle applies when adding a new reference from a URL. Interaction with the reference in the left pane is going to require knowing its name, so I think the user should provide it.